File: programming/THEDSA/Arrays/trappingRainWater.cpp
#include<bits/stdc++.h>
using namespace std;
int trap(vector<int>& height) {
        int res=0;
        for(int i=1;i<height.size()-1;i++){
            int left=height[i];
            for(int j=0;j<i;j++){
                left=max(left,height[j]);
            
            }int right=height[i];
            for(int j=i+1;j<height.size();j++){
                right=max(right,height[j]);
            }res=res + (min(left,right)-height[i]);
        }return res;
    }

int main(){
    vector<int>v={0,1,0,2,1,0,1,3,2,1,2,1};
    int res=trap(v);
    cout<<res<<endl;
}

File: programming/javaClasses/Arrays/TrappingRainWater.java
import java.util.*;
class TrappingRainWater{
    public int trap(int[] height) {
        if(height.length==1 || height.length==2) return 0;
        int rightMax[]=new int[height.length];
        int leftMax[]=new int[height.length];
        int maxi=Integer.MIN_VALUE;
        for(int i=0;i<height.length;i++){
            maxi=Math.max(maxi,height[i]);
            leftMax[i]=maxi;
        }
        maxi=Integer.MIN_VALUE;
        for(int i=height.length-1;i>=0;i--){
            maxi=Math.max(maxi,height[i]);
            rightMax[i]=maxi;
        }
        int res=0;
        for(int i=0;i<height.length;i++){
            res=res+Math.min(rightMax[i],leftMax[i])-height[i];
        }
        return res;
        
    }
    public static void main(String args[]){
        int arr[]={4,2,0,3,2,5};
        System.out.println(trap(arr));
    }
}